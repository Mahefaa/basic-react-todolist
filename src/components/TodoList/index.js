import {useState} from "react";
import Task from "../Task";
import List from "../List";
import style from "./index.module.css"

export default function TodoList() {
    const [list, setList] = useState([]);
    const [input, setInput] = useState("");
    const handleKeyDown = (e) => {
        if (e.key === "Enter" && input.trim().length !== 0) {
            setList(prevState => [...prevState, new Task(input)])
            setInput("");
        }
    }
    const handleChange = (e) => {
        setInput(e.target.value);
    }
    return (
        <div className={style.kanban__board}>
            <span className={style.input}>
                <input
                    type={"text"}
                    onKeyDown={handleKeyDown}
                    value={input}
                    onChange={handleChange}
                    placeholder={"What needs to be done ?"}
                    data-testid={"task-input"}
                />
            </span>
            <div className={style.list__container}>
                <List list={list} status={"TODO"} setList={setList}/>
                <List list={list} status={"DONE"} setList={setList}/>
            </div>
        </div>
    );
}