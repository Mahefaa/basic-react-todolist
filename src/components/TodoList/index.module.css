.container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
}

.input {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    width: 25vw;
    margin: 20px auto;
}

.list__container {
    flex-grow: 1;
    width: 100%;
    display: flex;
    flex-direction: row;
    height: 100%;
    justify-content: space-around;
}

.kanban__board {
    border: 2px solid black;
    min-width: 50vw;
    height: 80vh;
    background-color: rgba(232, 230, 230, 0.8);
    overflow: hidden;
}