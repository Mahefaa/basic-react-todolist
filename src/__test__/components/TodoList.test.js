import renderer from "react-test-renderer";
import TodoList from "../../components/TodoList";
import {cleanup, fireEvent, render, screen} from "@testing-library/react";

afterEach(cleanup);

test("TodoList component is rendered correctly", () => {
    const tree = renderer
        .create(<TodoList/>)
        .toJSON();

    expect(tree).toMatchSnapshot();
})

const LIST_TODO_TEST_ID = "list-TODO";
const LIST_DONE_TEST_ID = "list-DONE";
const FIRST_TODO_TEST_ID = "TODO-0";
const NEW_TODO_VALUE = "new todo";
const VALID_INPUT_KEYDOWN = "Enter";
const TASK_INPUT_TEST_ID = "task-input";
const INVALID_INPUT_KEYDOWN = "Space";
describe("Modify TodoList elements", () => {
    it("should add a new task with status Todo", () => {
        render(<TodoList/>)
        const input = screen.getByTestId(TASK_INPUT_TEST_ID);

        fireEvent.change(input, {target: {value: NEW_TODO_VALUE}});
        fireEvent.keyDown(input, {key: VALID_INPUT_KEYDOWN});
        const listOfTodos = screen.getByTestId(LIST_TODO_TEST_ID);

        expect(input).toHaveValue("");
        expect(listOfTodos.children.length).toEqual(1);
        expect(listOfTodos).toHaveTextContent(NEW_TODO_VALUE);
    })
    it("Input value should not be submitted if pressed VALID_INPUT_KEYDOWN isn't \"Enter\"", () => {
        render(<TodoList/>)
        const INPUT_TEST_VALUE = "Test value"
        const input = screen.getByTestId(TASK_INPUT_TEST_ID);

        fireEvent.change(input, {target: {value: INPUT_TEST_VALUE}});
        fireEvent.keyDown(input, {key: INVALID_INPUT_KEYDOWN});
        const list = screen.getByTestId(LIST_TODO_TEST_ID);

        expect(input).toHaveValue(INPUT_TEST_VALUE);
        expect(list.children.length).toEqual(0);
    })
    it("should move a task from todo to done", () => {
        render(<TodoList/>)
        const input = screen.getByTestId(TASK_INPUT_TEST_ID);

        fireEvent.change(input, {target: {value: NEW_TODO_VALUE}});
        fireEvent.keyDown(input, {key: VALID_INPUT_KEYDOWN});
        const checkBox = screen.getByTestId(FIRST_TODO_TEST_ID);
        fireEvent.click(checkBox);
        const listOfTodos = screen.getByTestId(LIST_TODO_TEST_ID);
        const listOfDone = screen.getByTestId(LIST_DONE_TEST_ID);

        expect(listOfTodos.children.length).toEqual(0);
        expect(listOfTodos).not.toHaveTextContent(NEW_TODO_VALUE)
        expect(listOfDone.children.length).toEqual(1);
        expect(listOfDone).toHaveTextContent(NEW_TODO_VALUE);
    })
})