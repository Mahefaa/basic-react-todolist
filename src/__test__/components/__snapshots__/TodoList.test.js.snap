// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component is rendered correctly 1`] = `
<div
  className="kanban__board"
>
  <span
    className="input"
  >
    <input
      data-testid="task-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="What needs to be done ?"
      type="text"
      value=""
    />
  </span>
  <div
    className="list__container"
  >
    <div
      className="list"
    >
      <h1
        className="title"
      >
        TODO
      </h1>
      <div
        className="list__items"
        data-testid="list-TODO"
      />
    </div>
    <div
      className="list"
    >
      <h1
        className="title"
      >
        DONE
      </h1>
      <div
        className="list__items"
        data-testid="list-DONE"
      />
    </div>
  </div>
</div>
`;
